[CONTENT] Newsletter 102 - 2022/2023 news (#117)
reactos/web-content: content/blogs/newsletter-102.md

While the [ReactOS Twitter account](https://twitter.com/reactos) does provide announcements, posts about working applications and such from time to time, much of what is happening with the project
as a whole isn't mentioned. Most of what we are going to talk about is the current situation with releases and overall ReactOS development.

## Why does it take so long to release a new version of ReactOS? Are you guys dead?
The latest release of ReactOS is 0.4.14, published on 16th of December, 2021. That release alone took a year to be released. Back then the ReactOS followed the 3-months cadence, which every 3 months
a new release would be released to the public. But since 2021, ReactOS is still at 0.4.14. Are you guys dead?

The answer is certainly no, the way we handle releases has changed. Back in the day the way ReactOS releases were made it was done for the sake of quantity than quality. Every new release after 3 months
is similar to each other, with exceptions of a very few features and bug fixes shipped to the release. But overall the net gap between the releases wasn't that big.

> Now there is a rule that for a development release to reach "Release Candidate" status, it has to have a reasonable amount of regressions (no more than 20) and that the stability isn't badly impacted

Reviewer: ```suggestion
Now there is a rule that for a development release to reach "Release" status, it has to have a reasonable amount of regressions (no more than 20) and that the stability isn't badly impacted
```
There is no special rule to reach Release-Candidate status. RC0 is a direct fork of the master branch without any changes.
RC1 has the first fix.
But the switch from Release Candidate to the first release (release-0), *that*'s when the "only 20 known regressions allowed" restriction kicks in.

So it is expected that it will take longer for a new release to be published, although such releases will now be substantial with lots of improvements than they were before.
If you're impatient to test the latest changes and features you can always try the [nightly builds](https://reactos.org/getbuilds) (also called experimental builds).

## Development status
In the last year and during the beginning of 2023, the ReactOS developers and contributors alike are working on many fronts of the project, the top focused area being the kernel.
Other areas that isn't kernel related are the applications, specifically our Paint and Notepad programs, the Input Method Editor (IME) as well as other stuff such as the ReactOS
testcase infrastructure.

## ReactOS x64 port status
The x64 port of ReactOS, while still in its infancy, has steadily gotten better in terms of stability. Back then ReactOS could barely boot and if it ever did, it usually resulted
in an immediate crash or a BSoD by just doing simple tasks. With the help of Timo Kreuzer (tkreuzer), the x64 port has gotten to a somewhat tolerable state where ReactOS can now

> operate without big problems. Although issues still do remain, for instance audio does not currently work in x64, more apps have started to work in one way or another.

Reviewer 2: That's not very accurate either. Both VMVideo and even VBoxVideo (with a small workaround) do in fact work on ReactOS x64. There was even a report of some NVidia x64 driver working. The issue is with the installer that breaks due to lack of WoW64 (at least in case of VBox). Let's just generalize and say that compatibility is currently inferior to x86 build and maybe also mention that x86 software still wouldn't run. Maybe something like this:

```suggestion
operate without big problems. Although issues still do remain and compatibility is inferior to the x86 version, more applications and even hardware drivers have started to work. Note that the x64 port currently will not run x86 applications due to lack of WoW64.
```

> operate without big problems. Although issues still do remain, for instance audio does not currently work in x64, more apps have started to work in one way or another.

Reviewer 3: @archeYR "Small workaround" means "things that are not in master" or "few hacks". AC97 is currently (and surprisingly) working under master. So... Let's say we are not compatible with x64 graphical drivers yet. The way you wrote it likes me more (less concrete). 

The same way I don't say that Intel graphic drivers works under Master x86, or the x64 version works under Real Hardware without touching anything in the code. That is simply a lie. And the people who read this must have clear the info about what can they use the x64 iso and what not.

> operate without big problems. Although issues still do remain, for instance audio does not currently work in x64, more apps have started to work in one way or another.

Reviewer 2: @julenuri
> "Small workaround" means "things that are not in master" or "few hacks". 

Not necessarily, it's just a registry change to lower the acceleration level (disable cursor acceleration). No out of tree changes are needed. If anyone is willing to install the VBox driver manually, they might as well apply the registry change in setup file which is very easy.  Apart from that, not everyone uses just VBox, VMware driver installs and works fine without any changes, and it even has OpenGL acceleration. To back my claim:
![Screenshot_20230414_204423](https://user-images.githubusercontent.com/32551254/232130737-e92c0aa0-67fa-4c15-8451-7bafd5505765.png)
> Let's say we are not compatible with x64 graphical drivers yet. 

But that's not true, some of the drivers do load and mostly work the same as if it was x86. 
> The way you wrote it likes me more (less concrete).

But it's true and gives a message to lower expectations from the x64 port in general, which is what we want in the end I guess.
> The same way I don't say that Intel graphic drivers works under Master x86, or the x64 version works under Real Hardware without touching anything in the code. That is simply a lie.

I did not say that it needs out of tree changes and I would never want to suggest anything like that in the official newsletter either...

> And the people who read this must have clear the info about what can they use the x64 iso and what not.

I think it is more important to put the correct info in the first place. I might as well say to keep the "audio doesn't work" because it indeed doesn't work in VMware on x64 (along USB and few others) and I don't care about other environments, but does it mean that audio really doesn't work at all? You seem to suggest otherwise with VBox/KVM. Also, if you've a feeling that more info is needed regarding the x64 port status and you know other specifics that apply in all cases (like missing WoW64), feel free to suggest them.

I don't think this needs to be discussed any further, but if you still want to, please move it somewhere else (ideally DM me (illen) on mattermost or matrix) to stop spamming this PR.

- The Client/Server Runtime Subsystem (CSRSS) has been revamped by Hermès BÉLUSCA - MAÏTO. This includes the refactor of the CSRSS codebase into a CSR library (`csrlib`), the CSR server (`csrsrv`) and the CSR executable subsystem itself (`csrss`). This allows ReactOS to use other subsystems other than the Win32 one ([#4802](https://github.com/reactos/reactos/pull/4802)).

For more updates and latest news, check our Twitter [@reactos](https://twitter.com/reactos).

## Beyond the main tree
ReactOS has always been more than just an operating system.
We are a community of people focused around the Windows ecosystem — with various projects to research and document Windows internals, improve Windows support in applications, or otherwise make life easier for the broader Windows developer community.
In fact, most developers likely had their first encounter with ReactOS not by downloading the operating system, but when looking up an undocumented API on the web and ending up in our [Doxygen-generated documentation](https://doxygen.reactos.org).

Consequently, much work within the past two years also happened outside our main source tree, and it's worthwhile to take a look at that:

> 

Reviewer: ```suggestion

Joachim Henze continued to improve all the older releases from 0.4.7 to 0.4.14 by porting back safe fixes for regressions and some optimizations, you can find updated isos for all of them on [sourceforge](https://sourceforge.net/projects/reactos/files/ReactOS/). For a complete list of the applied fixes you may have a look at [0.4.7](https://git.reactos.org/?p=reactos.git;a=shortlog;h=refs/heads/releases/0.4.7), [0.4.8](https://git.reactos.org/?p=reactos.git;a=shortlog;h=refs/heads/releases/0.4.8), [0.4.9](https://git.reactos.org/?p=reactos.git;a=shortlog;h=refs/heads/releases/0.4.9), [0.4.10](https://git.reactos.org/?p=reactos.git;a=shortlog;h=refs/heads/releases/0.4.10), [0.4.11](https://git.reactos.org/?p=reactos.git;a=shortlog;h=refs/heads/releases/0.4.11), [0.4.12](https://git.reactos.org/?p=reactos.git;a=shortlog;h=refs/heads/releases/0.4.12), [0.4.13](https://git.reactos.org/?p=reactos.git;a=shortlog;h=refs/heads/releases/0.4.13), [0.4.14](https://git.reactos.org/?p=reactos.git;a=shortlog;h=refs/heads/releases/0.4.14).
```

---
title:       "Newsletter 102 - 2022/2023 news"
author:      "GeoB99"
date:        2023-03-25
tags:        [ newsletter ]
---

> Hello ReactOS followers and enthusiasts! It has been quite a long time since we published the 101th newsletter and so far no further updates have been posted since then.

Reviewer 4: ```suggestion
Hello ReactOS followers and enthusiasts! It has been quite a long time since we published the [101th newsletter](/blogs/newsletter-101/) and so far no further updates have been posted since then.
```

While the [ReactOS Twitter account](https://twitter.com/reactos) does provide announcements, posts about working applications and such from time to time, much of what is happening with the project as a whole isn't mentioned. Most of what we are going to talk about is the current situation with releases and overall ReactOS development.

## Why does it take so long to release a new version of ReactOS? Are you guys dead?
The latest release of ReactOS is 0.4.14, published on 16th of December, 2021. That release alone took a year to be released. Back then the ReactOS followed the 3-months cadence, which every 3 months a new release would be released to the public. But since 2021, ReactOS is still at 0.4.14. Are you guys dead?

The answer is certainly no, the way we handle releases has changed. Back in the day the way ReactOS releases were made were done for the sake of quantity rather than quality. Every new release after 3 months is similar to the last one, with the exceptions of a small number of features and bug fixes added to the release. But overall the net gap between the releases wasn't that big.

> For a while now we have instated a rule that for a new version to reach "Release" status, it has to have a reasonable amount of regressions (no more than 20) and that the stability isn't badly impacted by the introduction of new features or code changes done during the development.In other words, the ReactOS project is now focusing on providing quality releases, which rules out the 3-months release cadence. This approach wasn't quite feasible for a development team with a size like ours anyway.

Reviewer 4: ```suggestion
For a while now we have instated a rule that for a new version to reach "Release" status, it has to have a reasonable amount of regressions (no more than 20) and that the stability isn't badly impacted by the introduction of new features or code changes done during the development.
In other words, the ReactOS project is now focusing on providing quality releases, which rules out the 3-months release cadence. This approach wasn't quite feasible for a development team with a size like ours anyway.
```

Reviewer 5: Without an extra newline, this won't show any different in the published page.

Reviewer 4: The problem is there are two sentences not even separated with space.

George Bișoc is currently working on improving the registry by implementing the necessary registry self-healing mechanisms and `CmCheckRegistry`, a fundamental function that validates the registry for any corrupted parts of the registry and takes appropriate actions at recovering a registry hive from such corrupt parts. While his work is currently in progress ([#4571](https://github.com/reactos/reactos/pull/4571)), so far he has received some positive results from testers. A small piece of his work has been merged, which fixes the registry flusher of ReactOS not working for almost a decade! This fix also solves the problem of the missing user profile due to forcefully shutting down the system, and as such you should no longer expect to see the infamous "Ctrl+Alt+Del" dialog prompt as often as before.

In addition he is working on the registry cache implementation ([#5088](https://github.com/reactos/reactos/pull/5088)), as ReactOS doesn't properly sync the cached information which leads to inconsistent results.

> Currently most of this work is still upheld because it would need some code refactoring in our bootloader, FreeLdr, as the implemented core registry healing routines makes bootloader size larger than what it should be which prevents the AMD64 build of ReactOS booting up due to a larger bootloader size.

Reviewer 4: ```suggestion
Currently most of this work is still being upheld because it makes our bootloader size larger than what it should be, preventing the AMD64 build of ReactOS booting up.
The implemented core registry healing routines require some refactoring in our bootloader, FreeLdr, before becoming mainstream.
```

---
title:       "Newsletter 102 - 2022/2023 news"
author:      "GeoB99"
date:        2023-03-25
tags:        [ newsletter ]
---

Hello ReactOS followers and enthusiasts! It has been quite a long time since we published the [101th newsletter](/blogs/newsletter-101/) and so far no further updates have been posted since then.
While the [ReactOS Twitter account](https://twitter.com/reactos) does provide announcements, posts about working applications and such from time to time, much of what is happening with the project as a whole isn't mentioned. Most of what we are going to talk about is the current situation with releases and overall ReactOS development.

## Why does it take so long to release a new version of ReactOS? Are you guys dead?

> The latest release of ReactOS is 0.4.14, published on 16th of December, 2021. That release alone took a year to be released. Back then the ReactOS followed the 3-months cadence, which every 3 months a new release would be released to the public. But since 2021, ReactOS is still at 0.4.14. Are you guys dead?

Reviewer 6: ```suggestion
The latest release of ReactOS is 0.4.14, published on 16th December, 2021. That release alone took a year to be engineered. Back then, ReactOS followed a 3-month cadence, releasing a new version every 3 months. But since 2021, ReactOS is still at 0.4.14. Are you guys dead?
```

Fixing grammar and not repeating the word "release" that often.

> The answer is certainly no, the way we handle releases has changed. Back in the day the way ReactOS releases were made were done for the sake of quantity rather than quality. Every new release after 3 months is similar to the last one, with the exceptions of a small number of features and bug fixes added to the release. But overall the net gap between the releases wasn't that big.

Reviewer 6: ```suggestion
The answer is certainly no, the way we handle releases has changed. Back in the day, ReactOS releases were made for the sake of quantity rather than quality. Every new release after 3 months was similar to the previous one, with the exception of a small number of added features and bug fixes. But overall, the net gap between the releases wasn't that big.
```

> For a while now we have instated a rule that for a new version to reach "Release" status, it has to have a reasonable amount of regressions (no more than 20) and that the stability isn't badly impacted by the introduction of new features or code changes done during the development.

Reviewer 6: ```suggestion
For a while now we have instated a rule that for a new version to reach "Release" status, it needs to have a reasonably low amount of regressions (no more than 20) and the stability mustn't be badly impacted by the introduction of new features or code changes done during development.
```

> In other words, the ReactOS project is now focusing on providing quality releases, which rules out the 3-months release cadence. This approach wasn't quite feasible for a development team with a size like ours anyway.

Reviewer 6: ```suggestion
In other words, the ReactOS project is now focusing on providing quality releases, which rules out the 3-month release cadence. That previous approach wasn't feasible for a development team with a size like ours anyway.
```

It is expected that it will take longer for a new release to be published, although these releases will now be much more substantial, with a much bigger number of improvements.
If you're impatient to test the latest changes and features you can always try the [nightly builds](https://reactos.org/getbuilds) (also called experimental builds).

## Development status
In the last year and during the beginning of 2023, the ReactOS developers and contributors alike are working on many parts of the project, the top focused area being the kernel.

> Other areas that aren't kernel related are the applications, specifically our Paint and Notepad programs, the Input Method Editor (IME) as well as other stuff such as the ReactOS testcase infrastructure.

Reviewer 6: ```suggestion
Other areas that aren't kernel-related are the applications, specifically our Paint and Notepad programs, the Input Method Editor (IME) as well as other stuff such as the ReactOS testing infrastructure.
```

## ReactOS x64 port status
The x64 port of ReactOS, while still in its infancy, has steadily gotten better in terms of stability. Back then ReactOS could barely boot and if it ever did, it usually resulted in an immediate crash or a BSoD by just doing simple tasks. With the help of Timo Kreuzer (tkreuzer), the x64 port has gotten to a somewhat tolerable state where ReactOS can now operate without big problems.

Although issues still do remain, and compatibility is inferior to the x86 version, more applications and even hardware drivers have started to work.
Note that the x64 port currently will not run **ANY** x86 applications due to lack of WoW64.

{{< gallery >}}
{{< figure link="/img/blogs/newsletter-102/vlc.png" src="/img/blogs/newsletter-102/vlc.png" >}}
{{< figure link="/img/blogs/newsletter-102/emule.png" src="/img/blogs/newsletter-102/emule.png" >}}
{{< figure link="/img/blogs/newsletter-102/sumatra.png" src="/img/blogs/newsletter-102/sumatra.png" >}}
{{< figure link="/img/blogs/newsletter-102/mixxxdj.png" src="/img/blogs/newsletter-102/mixxxdj.png" >}}
{{< /gallery >}}

## WHS test fixes

> The WHS test bot (acronym that stands for Windows Home Server) is a bot that runs Windows Server 2003 where some of its system files are exchanged with ReactOS' system files.

Reviewer 6: No, that's wrong. Windows Home Server is just a poor man's Windows Server 2003 (literally, it was 80€ in 2011).
We don't exchange any files by ReactOS files on that machine (such a testbot used to exist, but we're not running it at the moment).

testcase infrastructure.

## ReactOS x64 port status
The x64 port of ReactOS, while still in its infancy, has steadily gotten better in terms of stability. Back then ReactOS could barely boot and if it ever did, it usually resulted
in an immediate crash or a BSoD by just doing simple tasks. With the help of Timo Kreuzer (tkreuzer), the x64 port has gotten to a somewhat tolerable state where ReactOS can now
operate without big problems.

Although issues still do remain, and compatibility is inferior to the x86 version, more applications and even hardware drivers have started to work.
Note that the x64 port currently will not run x86 applications due to lack of WoW64.

{{< gallery >}}
{{< figure link="/img/blogs/newsletter-102/vlc.png" src="/img/blogs/newsletter-102/vlc.png" >}}
{{< figure link="/img/blogs/newsletter-102/emule.png" src="/img/blogs/newsletter-102/emule.png" >}}
{{< figure link="/img/blogs/newsletter-102/sumatra.png" src="/img/blogs/newsletter-102/sumatra.png" >}}
{{< figure link="/img/blogs/newsletter-102/mixxxdj.png" src="/img/blogs/newsletter-102/mixxxdj.png" >}}
{{< /gallery >}}

## WHS test fixes

> WHS testcase infrastructure was known to fail a lot of times due to tests not passing in WHS. On January of 2023 during the hackfest, Timo Kreuzer (tkreuzer), Mark Jansen (learn_more) and

Reviewer 7: I think that it's before this first sentence, that a 1-sentence-long explanation of what "WHS testcase" is. (See our MM discussion of yesterday.)

In worst cases, the system would still boot up but stuck in the login prompt dialog (aka the "Ctrl+Alt+Del" dialog) as the user profile has been corrupted or missing. The latter generally happens when one force shutdowns ReactOS after first clean install.

George Bișoc is currently working on improving the registry by implementing the necessary registry self-healing mechanisms and `CmCheckRegistry`, a fundamental function that validates the registry
for any corrupted parts of the registry and takes appropriate actions at recovering a registry hive from such corrupt parts. While his work is currently in progress ([#4571](https://github.com/reactos/reactos/pull/4571)), so far he has received some positive results from testers. A small piece of his work has been merged, which fixes the registry flusher of ReactOS not working for almost a decade! This fix also
solves the problem of the missing user profile due to force shutdown of the system, as such you should no longer expect the infamous "Ctrl+Alt+Del" dialog prompt.

> In addition he is working on the registry cache implementation ([#5088](https://github.com/reactos/reactos/pull/5088)), as ReactOS doesn't properly sync the cached information thus leading to inconsistent results.

Reviewer 7: No idea whether you want to mention that we have to wait for these improvements to be merged, because they are also needed in freeldr, and we need to change some ways of doing things in freeldr to make its size shrink beforehand.

Author: So according to your point of view, shall I delete that section? If so then outright I will do that.

Reviewer 7: No, not necessarily deleting the section, but just adding a caveat about/explanation why those improvements are not yet merged in master.

When IME message handling was implemented, the IME composition string window and candidate window were displayed correctly.

Video: https://www.youtube.com/watch?v=Dv58IV_vUxY

Note that Japanese input for Command Prompt and Rich Edit Control is not yet available. Also, several glitches and instabilities remain. Also, the ReactOS IMM is currently for older systems and does not support the newer TSF-compatible IMEs.
Defects and problems should be reported at https://github.com/katahiromz/ImeStudy/issues.

{{< gallery >}}
{{< figure link="/img/blogs/newsletter-102/ime.png" src="/img/blogs/newsletter-102/ime.png" caption="Using IME on Notepad in ReactOS" >}}
{{< /gallery >}}

## Other highlights
- Dmitry Borisov (disean) fixed a BSoD that was caused due to a shutdown with the IDE driver stack from Windows Server 2003.
- The Win32 backend kernel-mode subsystem component (`win32k.sys`) is steadily being improved with improvements and bug fixes by many people, notably Timo Kreuzer, James Tabor (jimtabor), Jerome Gardou (zefklop), Thamatip Chitpong (TAN-Gaming) and Katayama Hirofumi MZ.

> - The Client/Server Runtime Subsystem (CSRSS) has been revamped by Hermès Bélusca-Maïto. This includes the refactor of the CSRSS codebase into a CSR library (`csrlib`), the CSR server (`csrsrv`) and the CSR executable subsystem itself (`csrss`). This allows ReactOS to use other subsystems other than the Win32 one ([#4802](https://github.com/reactos/reactos/pull/4802)).

Reviewer 7: ```suggestion
- Some parts of the Session Manager SubSystem (SMSS) have been cleaned up and revamped by Hermès Bélusca-Maïto ([#4821](https://github.com/reactos/reactos/pull/4821)). This includes cleaning up old and deprecated code in the SM client library (`SMLIB`). Further, one API has been implemented in order to allow ReactOS to use other subsystems other than the Win32 one. In addition, the Client/Server Runtime Subsystem (CSRSS) has been slightly adapted for this purpose ([#4802](https://github.com/reactos/reactos/pull/4802)).
```

solves the problem of the missing user profile due to force shutdown of the system, as such you should no longer expect the infamous "Ctrl+Alt+Del" dialog prompt.

In addition he is working on the registry cache implementation ([#5088](https://github.com/reactos/reactos/pull/5088)), as ReactOS doesn't properly sync the cached information thus leading to inconsistent results.

## IME project
The Input Method Editor, sometimes abbreviated as IME, is a component that facilitates the typing of characters not originally present on the input devices by using a sequence of characters.
IME in ReactOS has been for the most part a stub, but thankfully Katayama Hirofumi MZ has extensively worked on IME.
This greatly helps with CJK support as well as allowing installation of custom IMEs for different locales.

> 
> Katayama wrote an article about his work, Japanese part of which was written using MZ-IME in ReactOS, here is the English part of it:

Reviewer 4: ```suggestion
Katayama wrote [an article](https://katahiromz.fc2.page/reactos-jpn-20230410/) about his work, Japanese part of which was written using MZ-IME in ReactOS, here is the English part of it:

```
Add link, move an empty line around.

The default procedure required the handling of several IME messages. The default IME window is an invisible window that is essential in IME messaging; the EDIT control is a so-called text box that required proper handling of some IME messages.
When IME message handling was implemented, the IME composition string window and candidate window were displayed correctly.

Video: https://www.youtube.com/watch?v=Dv58IV_vUxY

Note that Japanese input for Command Prompt and Rich Edit Control is not yet available. Also, several glitches and instabilities remain. Also, the ReactOS IMM is currently for older systems and does not support the newer TSF-compatible IMEs.
Defects and problems should be reported at https://github.com/katahiromz/ImeStudy/issues.

{{< gallery >}}
{{< figure link="/img/blogs/newsletter-102/ime.png" src="/img/blogs/newsletter-102/ime.png" caption="Using IME on Notepad in ReactOS" >}}
{{< /gallery >}}

## Other highlights
- Dmitry Borisov (disean) fixed a BSoD that was caused due to a shutdown with the IDE driver stack from Windows Server 2003.

> - The Win32 backend kernel-mode subsystem component (`Win32k.sys`) is steadily being improved with improvements and bug fixes by many people, notably Timo Kreuzer, James Tabor (jimtabor), Jerome Gardou (zefklop), Thamatip Chitpong (TAN-Gaming) and Katayama Hirofumi MZ.

Reviewer 4: ```suggestion
- The Win32 backend kernel-mode subsystem component (`win32k.sys`) is steadily being improved with improvements and bug fixes by many people, notably Timo Kreuzer, James Tabor (jimtabor), Jerome Gardou (zefklop), Thamatip Chitpong (TAN-Gaming) and Katayama Hirofumi MZ.
```

> - The Client/Server Runtime Subsystem (CSRSS) has been revamped by Hermès BÉLUSCA - MAÏTO. This includes the refactor of the CSRSS codebase into a CSR library (`csrlib`), the CSR server (`csrsrv`) and the CSR executable subsystem itself (`csrss`). This allows ReactOS to use other subsystems other than the Win32 one ([#4802](https://github.com/reactos/reactos/pull/4802)).

Reviewer 4: ```suggestion
- The Client/Server Runtime Subsystem (CSRSS) has been revamped by Hermès Bélusca-Maïto. This includes the refactor of the CSRSS codebase into a CSR library (`csrlib`), the CSR server (`csrsrv`) and the CSR executable subsystem itself (`csrss`). This allows ReactOS to use other subsystems other than the Win32 one ([#4802](https://github.com/reactos/reactos/pull/4802)).
```

---
title:       "Newsletter 102 - 2022/2023 news"
author:      "GeoB99"
date:        2023-03-25
tags:        [ newsletter ]
---

Hello ReactOS followers and enthusiasts! It has been quite a long time since we published the 101th newsletter and so far no further updates have been posted since then.
While the [ReactOS Twitter account](https://twitter.com/reactos) does provide announcements, posts about working applications and such from time to time, much of what is happening with the project
as a whole isn't mentioned. Most of what we are going to talk about is the current situation with releases and overall ReactOS development.

## Why does it take so long to release a new version of ReactOS? Are you guys dead?

> The latest release of ReactOS is 0.4.14, published on 16th of December, 2021. That release alone took a year to be released. Back then the ReactOS followed the 3-months cadence, which every 3 months

Reviewer 5: For these markdown files, it's better to keep one sentence per line (so applying suggestions is easier), and don't include artificial newlines (they will be stripped by hugo anyway).

Paragraphs will be inserted where there are multiple newlines.

Reviewer 5: ![image](https://github.com/reactos/web-content/assets/610685/23adbace-8ffb-4730-83ba-b66832e5cedf)

a new release would be released to the public. But since 2021, ReactOS is still at 0.4.14. Are you guys dead?

> The answer is certainly no, the way we handle releases has changed. Back in the day the way ReactOS releases were made it was done for the sake of quantity than quality. Every new release after 3 months

Reviewer 5: ```suggestion
The answer is certainly no, the way we handle releases has changed. Back in the day the way ReactOS releases were made were done for the sake of quantity rather than quality. Every new release after 3 months
```

> is similar to each other, with exceptions of a very few features and bug fixes shipped to the release. But overall the net gap between the releases wasn't that big.

Reviewer 5: ```suggestion
is similar to the last one, with the exceptions of a small number of features and bug fixes added to the release. But overall the net gap between the releases wasn't that big.
```

> Now there is a rule that for a development release to reach "Release" status, it has to have a reasonable amount of regressions (no more than 20) and that the stability isn't badly impacted

Reviewer 5: ```suggestion
For a while now we have instated a rule that for a new version to reach "Release" status, it has to have a reasonable amount of regressions (no more than 20) and that the stability isn't badly impacted
```

> by the introduction of new features or code changes done during the development. In other words, the ReactOS project is now focusing on providing quality releases of which the 3-months cadence cannot
> be applied anymore. Such approach wasn't quite feasible for a development team size like ours anyway.

Reviewer 5: ```suggestion
by the introduction of new features or code changes done during the development. In other words, the ReactOS project is now focusing on providing quality releases, which rules out the 3-months release cadence.
This approach wasn't quite feasible for a development team with a size like ours anyway.
```

> So it is expected that it will take longer for a new release to be published, although such releases will now be substantial with lots of improvements than they were before.
> If you're impatient to test the latest changes and features you can always try the [nightly builds](https://reactos.org/getbuilds) (also called experimental builds).

Reviewer 5: ```suggestion
It is expected that it will take longer for a new release to be published, although these releases will now be much more substantial, with a much bigger number of improvements.
If you're impatient to test the latest changes and features you can always try the [nightly builds](https://reactos.org/getbuilds) (also called experimental builds).
```

## Development status

> In the last year and during the beginning of 2023, the ReactOS developers and contributors alike are working on many fronts of the project, the top focused area being the kernel.
> Other areas that isn't kernel related are the applications, specifically our Paint and Notepad programs, the Input Method Editor (IME) as well as other stuff such as the ReactOS
> testcase infrastructure.

Reviewer 5: ```suggestion
In the last year and during the beginning of 2023, the ReactOS developers and contributors alike are working on many parts of the project, the top focused area being the kernel.
Other areas that aren't kernel related are the applications, specifically our Paint and Notepad programs, the Input Method Editor (IME) as well as other stuff such as the ReactOS
testcase infrastructure.
```

## ReactOS x64 port status
The x64 port of ReactOS, while still in its infancy, has steadily gotten better in terms of stability. Back then ReactOS could barely boot and if it ever did, it usually resulted
in an immediate crash or a BSoD by just doing simple tasks. With the help of Timo Kreuzer (tkreuzer), the x64 port has gotten to a somewhat tolerable state where ReactOS can now
operate without big problems.

Although issues still do remain, and compatibility is inferior to the x86 version, more applications and even hardware drivers have started to work.

> Note that the x64 port currently will not run x86 applications due to lack of WoW64.

Reviewer 5: ```suggestion
Note that the x64 port currently will not run **ANY** x86 applications due to lack of WoW64.
This means that most applications cannot be installed, because generally installers are x86 applications.
```

{{< gallery >}}
{{< figure link="/img/blogs/newsletter-102/vlc.png" src="/img/blogs/newsletter-102/vlc.png" >}}
{{< figure link="/img/blogs/newsletter-102/emule.png" src="/img/blogs/newsletter-102/emule.png" >}}
{{< figure link="/img/blogs/newsletter-102/sumatra.png" src="/img/blogs/newsletter-102/sumatra.png" >}}
{{< figure link="/img/blogs/newsletter-102/mixxxdj.png" src="/img/blogs/newsletter-102/mixxxdj.png" >}}
{{< /gallery >}}

## WHS test fixes
The WHS test bot (acronym that stands for Windows Home Server) is a bot that runs Windows Server 2003 where some of its system files are exchanged with ReactOS' system files.

> Within this test bot it runs the whole ReactOS testsuite infrastructure, where the test bot ensures that each testcase passes the test results on a Windows environment.

Reviewer 5: ```suggestion
Within this test bot it runs the whole ReactOS testsuite infrastructure, ensuring that each testcase passes the test results on a Windows environment.
```

This helps understanding how Windows behaves when certain test inputs are being given to it.

> WHS testcase infrastructure was known to fail a lot of times due to tests not passing in WHS. On January of 2023 during the hackfest, Timo Kreuzer (tkreuzer), Mark Jansen (learn_more) and

Reviewer 5: ```suggestion
WHS testcase infrastructure was known to fail a lot due to tests not passing, or even crashing. On January 2023 during the hackfest, Timo Kreuzer (tkreuzer), Mark Jansen (learn_more) and
```

> Thomas Faber (ThFabba) worked together to fix some of the testcases that prevented the WHS test bot to pass all the tests. Thanks for their efforts, WHS can now pass all the tests.

Reviewer 5: ```suggestion
Thomas Faber (ThFabba) worked together to fix some of the testcases. Thanks to their efforts, WHS can now pass all the tests.
```

was granted to anybody, regardless of the security restrictions imposed by security descriptors. This could lead to bad actors tamper and toy around with objects.

Thanks to his work, the hack in `SepAccessCheck` has been removed and security access checks are now operational and the kernel can prevent any unauthorized access from whoever tries
to access an object with improper access rights. His future plan for the Security Subsystem is to implement security auditing of all objects and miscellaneous improvements where needed.

## Registry improvements
For ReactOS to reach Beta phase it has to solve two major milestones -- stability and corruptions. Usually corruptions in ReactOS can manifest in form of filesystem and registry corruptions.
While filesystem corruptions can be prevented by choosing a more modern filesystem such as BtrFS which is what we currently support, registry corruptions are by themselves a death
sentence as ReactOS currently doesn't implement any self-healing mechanisms to recover the system from corrupted registry hives. As a result, the user is forced to do a clean reinstallation of the system.

> In worst cases, the system would still boot up but stuck in the login prompt dialog (aka the "Ctrl+Alt+Del" dialog) as the user profile has been corrupted or missing. The latter generally happens when one force shutdowns ReactOS after first clean install.

Reviewer 5: ```suggestion
Worst case, the system would still boot up but be stuck on the login prompt dialog (aka the "Ctrl+Alt+Del" dialog) as the user profile has been corrupted or missing. The latter generally happens after ReactOS is forcefully shutdown after a clean install.
```

George Bișoc is currently working on improving the registry by implementing the necessary registry self-healing mechanisms and `CmCheckRegistry`, a fundamental function that validates the registry
for any corrupted parts of the registry and takes appropriate actions at recovering a registry hive from such corrupt parts. While his work is currently in progress ([#4571](https://github.com/reactos/reactos/pull/4571)), so far he has received some positive results from testers. A small piece of his work has been merged, which fixes the registry flusher of ReactOS not working for almost a decade! This fix also

> solves the problem of the missing user profile due to force shutdown of the system, as such you should no longer expect the infamous "Ctrl+Alt+Del" dialog prompt.

Reviewer 5: ```suggestion
solves the problem of the missing user profile due to forcefully shutting down the system, and as such you should no longer expect to see the infamous "Ctrl+Alt+Del" dialog prompt as often as before.
```

> In addition he is working on the registry cache implementation ([#5088](https://github.com/reactos/reactos/pull/5088)), as ReactOS doesn't properly sync the cached information thus leading to inconsistent results.

Reviewer 5: ```suggestion
In addition he is working on the registry cache implementation ([#5088](https://github.com/reactos/reactos/pull/5088)), as ReactOS doesn't properly sync the cached information which leads to inconsistent results.
```

> Currently most of this work is still upheld because it would need some code refactoring in our bootloader, FreeLdr, as the implemented core registry fixup routines bloats the bootloader.

Reviewer 4: Bloat is something technically unneeded (like eye-candy stuff), so in this case it's not bloat, it just makes its size more than wanted:
```suggestion
Currently most of this work is still upheld because it would need some code refactoring in our bootloader, FreeLdr, as the implemented core registry healing routines makes bootloader size larger than we want it to be.
```

Reviewer 5: 'larger than we want it to be' --> It can't boot with a too big bootloader, so it is not really wanting, but a necessity to bring size down.

Reviewer 4: Yes, this would be more precise. :+1:

## IME project

> The Input Method Editor, sometimes abbreviated as IME, is a component that facilitates the typing of characters not originally present on the input devices by using a sequence of characters.
> IME in ReactOS has been for the most part a stub, but thankfully Katayama Hirofumi MZ has extensively worked on IME.
> This greatly helps with CJK support as well as allowing installation of custom IMEs for different locales.
> Katayama wrote [an article](https://katahiromz.fc2.page/reactos-jpn-20230410/) about his work, Japanese part of which was written using MZ-IME in ReactOS, here is the English part of it:

Reviewer 5: ```suggestion
The Input Method Editor (or IME) is a component that facilitates the typing of characters not originally present on the input devices by using a sequence of characters.
IME in ReactOS has been for the most part a stub, but thankfully Katayama Hirofumi MZ has extensively worked on it.
This greatly helps with CJK support, as well as allowing installation of custom IMEs for different locales.
Katayama wrote [an article](https://katahiromz.fc2.page/reactos-jpn-20230410/) about his work, the Japanese part of it was written using MZ-IME in ReactOS, here is the English part of it:
```

Seven months have passed since my last report, and I would like to explain the current status of Japanese input in ReactOS.
We have confirmed that Japanese ReactOS recognizes and can actually use MZ-IME Japanese input, a home-made test IME.

After the IMM32 implementation was almost complete, the remaining issue was the implementation on the USER32 side, where the default procedure and the processing of the default IME window and EDIT control had not yet been written.
The default procedure required the handling of several IME messages. The default IME window is an invisible window that is essential in IME messaging; the EDIT control is a so-called text box that required proper handling of some IME messages.
When IME message handling was implemented, the IME composition string window and candidate window were displayed correctly.

Video: https://www.youtube.com/watch?v=Dv58IV_vUxY

> Note that Japanese input for Command Prompt and Rich Edit Control is not yet available. Also, several glitches and instabilities remain. Also, the ReactOS IMM is currently for older systems and does not support the newer TSF-compatible IMEs.
> Defects and problems should be reported at https://github.com/katahiromz/ImeStudy/issues.

Reviewer 5: ```suggestion
Note that Japanese input for Command Prompt and Rich Edit Control is not yet available. Several glitches and instabilities remain. Please note that the ReactOS IMM is currently developed for older systems and does not support the newer TSF-compatible IMEs.
Defects and problems can be reported at https://github.com/katahiromz/ImeStudy/issues.
```

{{< gallery >}}
{{< figure link="/img/blogs/newsletter-102/ime.png" src="/img/blogs/newsletter-102/ime.png" caption="Using IME on Notepad in ReactOS" >}}
{{< /gallery >}}

## Other highlights
- Dmitry Borisov (disean) fixed a BSoD that was caused due to a shutdown with the IDE driver stack from Windows Server 2003.
- The Win32 backend kernel-mode subsystem component (`win32k.sys`) is steadily being improved with improvements and bug fixes by many people, notably Timo Kreuzer, James Tabor (jimtabor), Jerome Gardou (zefklop), Thamatip Chitpong (TAN-Gaming) and Katayama Hirofumi MZ.

> - Some parts of the Session Manager SubSystem (SMSS) have been cleaned up and revamped by Hermès Bélusca-Maïto ([#4821](https://github.com/reactos/reactos/pull/4821)). This includes cleaning up old and deprecated code in the SM client library (`SMLIB`). Further, one API has been implemented in order to allow ReactOS to use other subsystems other than the Win32 one. In addition, the Client/Server Runtime Subsystem (CSRSS) has been slightly adapted for this purpose ([#4802](https://github.com/reactos/reactos/pull/4802)).

Reviewer 5: ```suggestion
- Some parts of the Session Manager SubSystem (SMSS) have been cleaned up and revamped by Hermès Bélusca-Maïto ([#4821](https://github.com/reactos/reactos/pull/4821)). This includes cleaning up old and deprecated code in the SM client library (`SMLIB`). Furthermore, one API has been implemented in order to allow ReactOS to use other subsystems other than the Win32 one. In addition, the Client/Server Runtime Subsystem (CSRSS) has been slightly adapted for this purpose ([#4802](https://github.com/reactos/reactos/pull/4802)).
```
